File: JAVA/HeapSort.java
public class heapsort {
	
	public void sort(int arr[], int n) {
		for(int i=n/2-1; i>=0; i--) {
			maxHeapify(arr, n, i);
		}
		
		for(int i=n-1; i>0; i--) {
			int temp = arr[0];
			arr[0] = arr[i];
			arr[i] = temp;
			maxHeapify(arr, i, 0);
		}
	}
	
	public void maxHeapify(int arr[], int n, int i) {
		int largest = i;
		int l = 2*i + 1;
		int r = 2*i + 2;
		
		if(l<n && arr[largest] < arr[l]) {
			largest = l;
		}
		
		if(r<n && arr[largest] < arr[r]) {
			largest = r;
		}
		
		if(largest != i) {
			int temp = arr[largest];
			arr[largest] = arr[i];
			arr[i] = temp;
			
			maxHeapify(arr, n, largest);
		}
	}
	
	public static void main(String[] args) {
		int arr[] = { 12, 11, 13, 5, 6, 7};
		
		heapsort obj = new heapsort();
		obj.sort(arr, 6);
		for(int element: arr) {
			System.out.print(element + " ");
		}
	}
	
}

File: JAVASCRIPT/heap_sort.js
// heap sort code in javascript

// creating a heap
class MaxHeap {
    constructor() {
        this.heap = [];
    }

    parentIndex(index) {
        return Math.floor((index - 1) / 2);
    }

    leftChildIndex(index) {
        return (2 * index + 1);
    }

    rightChildIndex(index) {
        return (2 * index + 2);
    }

    // to swap to elements
    swap(a, b) {
        let temp = this.heap[a];
        this.heap[a] = this.heap[b];
        this.heap[b] = temp;
    }

    // to insert element in heap
    insert(item) {
        this.heap.push(item);
        var index = this.heap.length - 1;
        var parent = this.parentIndex(index);
        while (this.heap[parent] && this.heap[parent] < this.heap[index]) {
            this.swap(parent, index);
            index = this.parentIndex(index);
            parent = this.parentIndex(index);
        }
    }

    // to delete element from heap
    delete_ele() {
        var item = this.heap.shift();
        this.heap.unshift(this.heap.pop());
        var index = 0;
        var leftChild = this.leftChildIndex(index);
        var rightChild = this.rightChildIndex(index);
        while (this.heap[leftChild] && this.heap[leftChild] > this.heap[index] || this.heap[rightChild] > this.heap[index]) {
            var max = leftChild;
            if (this.heap[rightChild] && this.heap[rightChild] > this.heap[max]) {
                max = rightChild
            }
            this.swap(max, index);
            index = max;
            leftChild = this.leftChildIndex(max);
            rightChild = this.rightChildIndex(max);
        }
        return item;
    }
}

// heap sort function
function heapSort(arr) {
    var sorted = [];
    var heap1 = new MaxHeap();

    for (let i = 0; i < arr.length; i++) {
        heap1.insert(arr[i]);
    }

    for (let i = 0; i < arr.length; i++) {
        sorted.push(heap1.delete_ele());
    }
    return sorted;
}

// array to be sorted
let array = [99, 55, 66, 22, 33, 77, 88, 11, 44]

var sortedArray = heapSort(array);
console.log(sortedArray);